examples/gno.land/p/demo/timelock/timelock.gno

```diff
@@ -0,0 +1,181 @@
+// Package timelock provides a library for scheduling, cancelling, and
+// executing time-locked operations in Gno. It ensures that
+// operations are only carried out after a specified delay and offers
+// mechanisms for managing and verifying the status of these operations.
+// This package leverages an AVL tree for efficient management of timestamps
+// and integrates role-based access control for administrative tasks.
+//
+// Example Usage:
+//
+//  import "gno.land/p/demo/timelock"
+//  import "gno.land/p/demo/accesscontrol"
+//
+//  Initialize timelock utility with an AVL tree and access control.
+//  timestamps := avl.NewTree()
+//  adminRole := accesscontrol.NewRoleData("admin", std.Address("admin-address"))
+//  timeLockUtil := timelock.NewTimeLockUtil(timestamps, adminRole, 30)
+//
+//  Schedule an operation with a delay of 60 seconds.
+//  id := seqid.ID()
+//  timeLockUtil.Schedule(id, 60)
+//
+//  Check if an operation is pending.
+//  isPending := timeLockUtil.IsPending(id)
+//
+//  Execute the operation when it is pending.
+//  if timeLockUtil.IsPending(id) {
+//      timeLockUtil.Execute(id)
+//  }
+//
+//  Update the minimum delay for future operations.
+//  timeLockUtil.UpdateDelay(45)
+
+package timelock // import "gno.land/p/demo/timelock"
+
+import (
+	"std"
+	"strconv"
+	"time"
+
+	"gno.land/p/demo/ufmt"
+
+	"gno.land/p/demo/accesscontrol"
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/seqid"
+)
+
+// Marks operations completed
+const DoneTimestamp = 1
+
+// Represents the status of a planned operation
+type OperationState int
+
+const (
+	Unset OperationState = iota
+	Pending
+	Ready
+	Done
+)
+
+// TimeLockUtil stores the necessary parameters for the timelock operations
+type TimeLockUtil struct {
+	timestamps    *avl.Tree
+	accessControl *accesscontrol.RoleData
+	minDelay      uint64
+}
+
+// New instance of TimeLockUtil
+func MustNewTimeLockUtil(timestamps *avl.Tree, accessControl *accesscontrol.RoleData, minDelay uint64) (*TimeLockUtil, error) {
+	if timestamps == nil || accessControl == nil {
+		return nil, ufmt.Errorf("timestamps and accesscontrol values must be different from nil")
```

Comment:
Perhaps `errors.New` would be better to use here; no formatting is happening.

Reply:
- https://github.com/gnolang/gno/pull/2307/commits/8da6e065c93489c876ae2ca331fb774f5937f758